src/plugins/mf_priority.cpp

```diff
@@ -554,11 +557,6 @@ static int validate_cb (flux_plugin_t *p,
     cur_active_jobs = bank_it->second.cur_active_jobs;
     max_active_jobs = bank_it->second.max_active_jobs;
 
-    // if a user's fairshare value is 0, that means they shouldn't be able
```

Comment:
The commit message still has me a little bit confused, though I'm sure it is just me. However, for my own edification, I'd like to understand. The commit message says:

>  The multi-factor priority plugin currently rejects jobs from a user
> who has a fairshare value of 0.0. To align this behavior more closely with
> the current implementation on our systems, the plugin should instead assign
> the lowest job priority for the job submitted.

However, the change as I see it in the code is that the minimum priority is now returned if the user's fairshare is _less than_ zero, not equal to zero. Jobs with `0` fairshare are now no longer rejected, but they don't get assigned `FLUX_JOB_PRIORITY_MIN` - instead they appear to get whatever priority is calculated in `priority_calculation()` taking into account the fairshare value of `0.`.

Is that the intended operation? If so maybe rewording the commit message would help.

Reply:
Ohhh, no it's not you at all, you are absolutely right. Right now the change I proposed is checking that a user's fairshare is _less than_ zero, not _less than or equal to 0_; the latter is the behavior I am trying to add in. I suppose I should change this comparison to

```c++
if (fshare_factor <= 0)
    return FLUX_JOB_PRIORITY_MIN;
```

So now if a user's fairshare is `0.0`, it should return a minimum priority. I should have noticed that myself, sorry about that. I can push up a fix.

Reply:
Ok, that makes more sense! Thanks

Reply:
I'm sorry, I meant to ping you that I pushed up a fix. I was waiting for Codecov to publish the report so that I could check that the lines were actually covered, but it's slower than I thought. :/

Reply:
@dongahn, I knew that testing for equality was not safe but here the code is testing `<= 0.` which I had considered "safe". Please advise if my assumption is incorrect. @cmoussa1, I guess one thing to consider is if you want to treat a very small value _epsilon_ as equivalent to zero, then you might want to use

```c
    if (fshare_factor < std::numeric_limits<double>::epsilon ())
        return FLUX_JOB_PRIORITY_MIN;
```

I'm not 100% sure if that is necessary in this case.

Reply:
>Oh, nice, thanks for pointing this out! In the context of this PR, I am comparing a double to an int instead of comparing two doubles. I'm guessing the precautions you pasted above also apply to this scenario? 

I guess so. The integer will be promoted to `double` for comparison...

Reply:
Can `fshare_factor` become a negative?

Reply:
Not to my knowledge, no. `fshare_factor` should be in the range `0.0 <= fshare_factor <= 1.0`

Reply:
So is this comparison just a sanity check that you won't have a negative`fshare_factor`? Or are you trying to set some threshold on a very small factor value?

Reply:
Ah, I apologize, some of my comments may have been confusing then since I assumed `fshare_factor` < 0 was a possibility.

Reply:
The comparison `fshare_factor <= 0` was put in place in the case that `fshare_factor` was `0.0`, a specific job priority `FLUX_JOB_PRIORITY_MIN` should be returned. I hadn't used `==` for comparison since I had a little knowledge (but definitely not extensive) of the dangers of comparing double values using `==`, so that's why I had used `<=`. 

So maybe yours and @grondo's suggestion of using `std::numeric_limits<double>::epsilon ()` should be used instead?

Reply:
@cmoussa1: From the code comment:

```
 * fairshare: the ratio between the amount of resources allocated vs. resources
 *     consumed.
 ```

Can `fshare_factor` really become zero? Probably just closer and closer to zero but never become zero since this is a ratio with a positive nominator? 

In any case, what is the criteria in determining the threshold to return `FLUX_JOB_PRIORITY_MIN` with? Why should that be `std::numeric_limits<double>::epsilon ()`?

Reply:
Ohhh, now that I'm typing this out, I _do_ briefly remember an offline conversation you and I had that `fshare_factor` could never really become zero, it could just get closer and closer to zero; I think I had an experimental branch where I confirmed it too. I'm sorry I didn't remember that!

Reply:
> In any case, what is the criteria in determining the threshold to return `FLUX_JOB_PRIORITY_MIN` with? Why should that be `std::numeric_limits<double>::epsilon ()`?

If I am understanding your question correctly, the criteria for returning `FLUX_JOB_PRIORITY_MIN` was if `fshare_factor` was truly `0.0`, but, if `fshare_factor` can never really become zero, then perhaps it is not necessary to check for this in the first place. Sorry if I misunderstood though.

Reply:
@cmoussa1: I was just trying to understand your algorithm a bit. What is the purpose of `FLUX_JOB_PRIORITY_MIN`? Do you have to signal the upper layer about a certain condition? If that condition never happens, it seems that the check is defeating the purpose. But if this is a condition that can happen, I thought I could be a sounding board for you. 

BTW I also see

```c
    priority = round ((fshare_weight * fshare_factor) +
                      (queue_weight * queue_factor) +
                      (urgency - 16));

    if (priority < 0)
        return FLUX_JOB_PRIORITY_MIN;
```

Can `priority` become zero here as well?

I don't think I understand the entire picture. Sorry if this is confusing you.

Reply:
> What is the purpose of FLUX_JOB_PRIORITY_MIN?

This is an integer priority (0) which signals that the job should not exit out of PRIORITY state and, as a result, will be held until it receives an integer priority > 0. This was supposed to be the priority returned for a job submitted by a user whose fairshare value was `0.0`, but I think we've since determined such condition is not possible, since fairshare can only get really really close to `0.0`.

> Can `priority` become zero here as well?

Yes, I think technically `priority` can become zero or less than zero. For example, if the weight for both factors (`fshare_weight` and `queue_weight`) are both 0, and the `urgency` for that job was < 16, then `priority` would be negative. But since we don't want to return a negative priority, we just return `FLUX_JOB_PRIORITY_MIN`. I think this is a very unlikely scenario however, as by default we would want the weights for both factors (`fshare_weight` and `queue_weight`) to be large positive integers, and I think only a sys admin would have the privilege to set an urgency < 16 for a job. I hope this makes sense.

Reply:
All make sense. I think you could either just not do that floating point comparison or have a configurable (or set) threshold, `t` such that if `fshare_factor` becomes less than that value, we will deem `FLUX_JOB_PROIRTY_MIN`. It isn't clear why that threshold to be an epsilon, though. But since this is just a policy choice, I wouldn't object to the use of the machine eepsilon as your pollicy for now, though.

Reply:
I think I'll just drop the floating point comparison for now. That seems like a reasonable choice. Thanks @dongahn for working through this with me and helping me reason it out, it was really helpful.